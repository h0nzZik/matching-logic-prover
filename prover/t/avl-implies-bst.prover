imports "t/definitions/avl.prover"
imports "t/definitions/bst.prover"

claim \implies( \and( avl( variable("H") { ArrayIntInt }
                         , variable("X") { Int }
                         , variable("F") { SetInt }
                         , variable("MIN") { Int }
                         , variable("MAX") { Int }
                         , variable("Height") { Int }
                         , variable("Balance") { Int }
                         )
                    )
              , \and( bst( variable("H") { ArrayIntInt }
                         , variable("X") { Int }
                         , variable("F") { SetInt }
                         , variable("MIN") { Int }
                         , variable("MAX") { Int }
                         )
                    )
              )
strategy search-bound(3)
//strategy simplify ; kt ; simplify
//       ; ( (right-unfold-Nth(0,0) ; simplify ; instantiate-existentials ; smt )
//         | (right-unfold-Nth(0,1) ; simplify ; instantiate-existentials ; smt )
//         | (instantiate-existentials ; smt )
//         | (right-unfold-Nth(0,2) ; simplify ; instantiate-existentials ; smt )
//         | noop
//         )
