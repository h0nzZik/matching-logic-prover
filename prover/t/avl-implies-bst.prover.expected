<prover>
  <k>
    .
  </k>
  <goals>
    <goal>
      <id>
        root
      </id>
      <active>
        true
      </active>
      <parent>
        .
      </parent>
      <claim>
        \implies ( \and ( avl ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , variable ( "Height" ) { Int } , variable ( "Balance" ) { Int } , .Patterns ) , .Patterns ) , \exists { .Patterns } \and ( bst ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) , .Patterns ) )
      </claim>
      <strategy>
        success
      </strategy>
      <trace>
        .
      </trace>
    </goal>
  </goals>
  <declarations>
    <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , .Patterns } \iff-lfp ( bst ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , 0 ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { .Patterns } \and ( gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) , 0 ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 2 , .Patterns ) , .Patterns ) , 0 ) , \equals ( variable ( "MIN" ) { Int } , variable ( "X" ) { Int } ) , \equals ( variable ( "MAX" ) { Int } , variable ( "X" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) ) , .Patterns ) , \exists { variable ( "F_J1" ) { SetInt } , variable ( "F_J2" ) { SetInt } , variable ( "MAX_L1" ) { Int } , variable ( "MAX_L2" ) { Int } , variable ( "MIN_K1" ) { Int } , variable ( "MIN_K2" ) { Int } , variable ( "X_I1" ) { Int } , variable ( "X_I2" ) { Int } , .Patterns } \and ( bst ( variable ( "H" ) { ArrayIntInt } , variable ( "X_I1" ) { Int } , variable ( "F_J1" ) { SetInt } , variable ( "MIN_K1" ) { Int } , variable ( "MAX_L1" ) { Int } , .Patterns ) , bst ( variable ( "H" ) { ArrayIntInt } , variable ( "X_I2" ) { Int } , variable ( "F_J2" ) { SetInt } , variable ( "MIN_K2" ) { Int } , variable ( "MAX_L2" ) { Int } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) , variable ( "X_I1" ) { Int } ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 2 , .Patterns ) , .Patterns ) , variable ( "X_I2" ) { Int } ) , gt ( variable ( "X" ) { Int } , variable ( "MAX_L1" ) { Int } , .Patterns ) , gt ( variable ( "MIN_K2" ) { Int } , variable ( "X" ) { Int } , .Patterns ) , \equals ( variable ( "MIN_K1" ) { Int } , variable ( "MIN" ) { Int } ) , \equals ( variable ( "MAX_L2" ) { Int } , variable ( "MAX" ) { Int } ) , \not ( isMember ( variable ( "X" ) { Int } , variable ( "F_J1" ) { SetInt } , .Patterns ) ) , \not ( isMember ( variable ( "X" ) { Int } , variable ( "F_J2" ) { SetInt } , .Patterns ) ) , \equals ( variable ( "F" ) { SetInt } , union ( singleton ( variable ( "X" ) { Int } , .Patterns ) , union ( variable ( "F_J1" ) { SetInt } , variable ( "F_J2" ) { SetInt } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F_J1" ) { SetInt } , variable ( "F_J2" ) { SetInt } , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , variable ( "Height" ) { Int } , variable ( "Balance" ) { Int } , .Patterns } \iff-lfp ( avl ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , variable ( "Height" ) { Int } , variable ( "Balance" ) { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , 0 ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , \equals ( variable ( "Balance" ) { Int } , 0 ) , \equals ( variable ( "Height" ) { Int } , 0 ) , .Patterns ) , \exists { .Patterns } \and ( gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) , 0 ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 2 , .Patterns ) , .Patterns ) , 0 ) , \equals ( variable ( "MIN" ) { Int } , variable ( "X" ) { Int } ) , \equals ( variable ( "MAX" ) { Int } , variable ( "X" ) { Int } ) , \equals ( variable ( "Balance" ) { Int } , 0 ) , \equals ( variable ( "Height" ) { Int } , 1 ) , \equals ( variable ( "F" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) ) , .Patterns ) , \exists { variable ( "B_1" ) { Int } , variable ( "F_1" ) { SetInt } , variable ( "H_1" ) { Int } , variable ( "MAX_1" ) { Int } , variable ( "MIN_1" ) { Int } , variable ( "X_1" ) { Int } , .Patterns } \and ( avl ( variable ( "H" ) { ArrayIntInt } , variable ( "X_1" ) { Int } , variable ( "F_1" ) { SetInt } , variable ( "MIN_1" ) { Int } , variable ( "MAX_1" ) { Int } , variable ( "H_1" ) { Int } , variable ( "B_1" ) { Int } , .Patterns ) , avl ( variable ( "H" ) { ArrayIntInt } , variable ( "X_1" ) { Int } , variable ( "F_1" ) { SetInt } , variable ( "MIN_1" ) { Int } , variable ( "MAX_1" ) { Int } , variable ( "H_1" ) { Int } , variable ( "B_1" ) { Int } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( variable ( "Balance" ) { Int } , minus ( variable ( "H_1" ) { Int } , variable ( "H_1" ) { Int } , .Patterns ) ) , gt ( variable ( "Balance" ) { Int } , -2 , .Patterns ) , gt ( 2 , variable ( "Balance" ) { Int } , .Patterns ) , \equals ( variable ( "Height" ) { Int } , plus ( max ( variable ( "H_1" ) { Int } , variable ( "H_1" ) { Int } , .Patterns ) , 1 , .Patterns ) ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) , variable ( "X_1" ) { Int } ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 2 , .Patterns ) , .Patterns ) , variable ( "X_1" ) { Int } ) , gt ( variable ( "X" ) { Int } , variable ( "MAX_1" ) { Int } , .Patterns ) , gt ( variable ( "MIN_1" ) { Int } , variable ( "X" ) { Int } , .Patterns ) , \equals ( variable ( "MIN_1" ) { Int } , variable ( "MIN" ) { Int } ) , \equals ( variable ( "MAX_1" ) { Int } , variable ( "MAX" ) { Int } ) , \not ( isMember ( variable ( "X" ) { Int } , variable ( "F_1" ) { SetInt } , .Patterns ) ) , \not ( isMember ( variable ( "X" ) { Int } , variable ( "F_1" ) { SetInt } , .Patterns ) ) , \equals ( variable ( "F" ) { SetInt } , union ( singleton ( variable ( "X" ) { Int } , .Patterns ) , union ( variable ( "F_1" ) { SetInt } , variable ( "F_1" ) { SetInt } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F_1" ) { SetInt } , variable ( "F_1" ) { SetInt } , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      symbol avl ( ArrayIntInt , Int , SetInt , Int , Int , Int , Int , .Sorts ) : Bool
    </declaration> <declaration>
      symbol bst ( ArrayIntInt , Int , SetInt , Int , Int , .Sorts ) : Bool
    </declaration>
  </declarations>
</prover>
